crates/holochain_state/src/source_chain.rs

```diff
@@ -1713,52 +1707,108 @@ pub mod tests {
             let action = chain
                 .put_weightless(action_builder, Some(entry), ChainTopOrdering::default())
                 .await?;
-
             chain.flush(&mock).await.unwrap();
 
             (action, entry_hash)
         };
 
+        // alice as chain owner should be unaffected by updates
+        // chain author grant should always be returned
+        assert_eq!(
+            chain
+                .valid_cap_grant(function.clone(), alice.clone(), secret.clone())
+                .await?,
+            Some(CapGrant::ChainAuthor(alice.clone())),
+        );
+        assert_eq!(
+            chain
+                .valid_cap_grant(function.clone(), alice.clone(), updated_secret.clone())
+                .await?,
+            Some(CapGrant::ChainAuthor(alice.clone())),
+        );
+
+        // bob must not get a valid cap grant with the initial cap secret,
+        // as it is invalidated by the cap grant update
+        assert_eq!(
+            chain
+                .valid_cap_grant(function.clone(), bob.clone(), secret.clone())
+                .await?,
+            None
+        );
+        // when bob provides the updated secret, validation should succeed,
+        // bob being an assignee
+        assert_eq!(
+            chain
+                .valid_cap_grant(function.clone(), bob.clone(), updated_secret.clone())
+                .await?,
+            Some(updated_grant.clone().into())
+        );
+
+        // carol must not get a valid cap grant even with the updated secret,
+        // as carol is not an assignee
+        assert_eq!(
+            chain
+                .valid_cap_grant(function.clone(), carol.clone(), updated_secret.clone())
+                .await?,
+            None
+        );
```

Comment:
```suggestion
        // carol must not get a valid cap grant with either the original secret (because it was replaced)
        // or the updated secret (because she is not an assignee)
        assert_eq!(
            chain
                .valid_cap_grant(function.clone(), carol.clone(), secret.clone())
                .await?,
            None
        );
        assert_eq!(
            chain
                .valid_cap_grant(function.clone(), carol.clone(), updated_secret.clone())
                .await?,
            None
        );
```